pmd-core/src/main/java/net/sourceforge/pmd/lang/ast/QualifiedName.java

```diff
@@ -0,0 +1,36 @@
+/**
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.ast;
+
+/**
+ * Basic interface for qualified names. That's used in the metrics framework to refer unambiguously to operations or
+ * classes. Language specific pmd modules should have at most one implementation of this interface, to allow safe
+ * downcasting from QualifiedName to e.g. JavaQualifiedName.
```

Comment:
See in [`ProjectMirror`](https://github.com/oowekyala/pmd/blob/metrics-abstraction/pmd-core/src/main/java/net/sourceforge/pmd/lang/metrics/ProjectMirror.java#L31), the methods defined here have to use the general interface QualifiedName as arguments. Since we can't implement these methods with a more specific type for the argument, we have to downcast it [here](https://github.com/oowekyala/pmd/blob/metrics-abstraction/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/metrics/PackageStats.java#L163) to use the language specific features of the QualifiedName implementation